### PR TITLE
crimson: add support for backfill, part 2 -- MOSDPGBackfillRemove

### PR DESCRIPTION
**Pull request by rzarzynski**
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/crimson/osd/recovery_backend.cc [hunk 1]
```diff
@@ -136,6 +136,25 @@ seastar::future<> RecoveryBackend::handle_backfill(
   }
 }
 
+seastar::future<> RecoveryBackend::handle_backfill_remove(
+  MOSDPGBackfillRemove& m)
+{
+  logger().debug("{} m.ls=", __func__, m.ls);
+  ceph_assert(m.get_type() == MSG_OSD_PG_BACKFILL_REMOVE);
```
Comment (tchaikov) [inline src/crimson/osd/recovery_backend.cc:143]: i'd suggest use `assert()` for smaller executable and less runtime overhead as i think this assert is mostly for document purpose.

File: src/crimson/osd/recovery_backend.cc [hunk 2]
```diff
@@ -136,6 +136,25 @@ seastar::future<> RecoveryBackend::handle_backfill(
...
+
+  ObjectStore::Transaction t;
+  for ([[maybe_unused]] const auto& [ soid, ver ] : m.ls) {
```
Comment (tchaikov) [inline src/crimson/osd/recovery_backend.cc:146]: nit, i'd suggest remove the spaces around `soid, ver`.

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/recovery_backend.cc (hunk 1 above)
Comment (rzarzynski) [reply src/crimson/osd/recovery_backend.cc]: Yeah, that exactly was its purpose. `assert` is better for this kind of work. Thanks!

**Review by tchaikov — APPROVED** [review]

File: src/crimson/osd/recovery_backend.cc [hunk 3]
```diff
@@ -136,6 +136,25 @@ seastar::future<> RecoveryBackend::handle_backfill(
   }
 }
 
+seastar::future<> RecoveryBackend::handle_backfill_remove(
+  MOSDPGBackfillRemove& m)
+{
+  logger().debug("{} m.ls=", __func__, m.ls);
```
Comment (ronen-fr) [inline src/crimson/osd/recovery_backend.cc:142]: This seems wrong... missing {}?

File: src/crimson/osd/backfill_state.h [hunk 4]
```diff
@@ -298,6 +298,8 @@ struct BackfillState::BackfillListener {
     const hobject_t& obj,
     const eversion_t& v) = 0;
 
+  virtual void maybe_flush() = 0;
```
Comment (ronen-fr) [inline src/crimson/osd/backfill_state.h:301]: Adding a comment re why termed "maybe"?

File: src/crimson/osd/recovery_backend.cc (hunk 3 above)
Comment (tchaikov) [reply src/crimson/osd/recovery_backend.cc:142]: addressed by https://github.com/ceph/ceph/pull/38501/commits/027b3037198b1cf58cbb919091f27ea2d86140bc

Comment (rzarzynski) [reply src/crimson/osd/recovery_backend.cc:142]: :+1:

File: src/crimson/osd/backfill_state.h (hunk 4 above)
Comment (rzarzynski) [reply src/crimson/osd/backfill_state.h:301]: *maybe* means to me the decision whether to flush-or-not belongs to an implementation; the interface doesn't enforce actual *flushing* on each call. I believed it's a common interpretation of such word usage; if you think it's different, I will comment on that.

On the other hand, this extra flexibility isn't currently used. Both implementations are always flushing, so perhaps we should rename to just `flush()`.

**Pull request merged**